File: src/main/java/org/jcodec/containers/dpx/DPXMetadata.java
package org.jcodec.containers.dpx;

import static java.lang.String.format;

public class DPXMetadata {
    public static final String V2 = "V2.0";
    public static final String V1 = "V1.0";

    public FileHeader file;
    public ImageHeader image;
    public ImageSourceHeader imageSource;
    public FilmHeader film;
    public TelevisionHeader television;
    public String userId;

    private static String smpteTC(int tcsmpte, boolean prevent_dropframe) {
        int ff = bcd2uint(tcsmpte & 0x3f);    // 6-bit hours
        int ss = bcd2uint(tcsmpte >> 8 & 0x7f);    // 7-bit minutes
        int mm = bcd2uint(tcsmpte >> 16 & 0x7f);    // 7-bit seconds
        int hh = bcd2uint(tcsmpte >> 24 & 0x3f);    // 6-bit frames
        boolean drop = (tcsmpte & 1 << 30) > 0L && !prevent_dropframe;  // 1-bit drop if not arbitrary bit
        return format("%02d:%02d:%02d%c%02d", hh, mm, ss, drop ? ';' : ':', ff);
    }

    private static int bcd2uint(int bcd) {
        int low = bcd & 0xf;
        int high = bcd >> 4;
        if (low > 9 || high > 9)
            return 0;
        return low + 10 * high;
    }

    public String getTimecodeString() {
        return smpteTC(television.timecode, false);
    }

}

File: src/main/java/org/jcodec/containers/dpx/DPXMetadata.kt
package org.jcodec.containers.dpx

import org.jcodec.common.StringUtils

class DPXMetadata {
    @JvmField
    var file: FileHeader? = null
    @JvmField
    var image: ImageHeader? = null
    @JvmField
    var imageSource: ImageSourceHeader? = null
    @JvmField
    var film: FilmHeader? = null
    @JvmField
    var television: TelevisionHeader? = null
    @JvmField
    var userId: String? = null
    val timecodeString: String
        get() = smpteTC(television!!.timecode, false)

    companion object {
        const val V2 = "V2.0"
        const val V1 = "V1.0"
        private fun smpteTC(tcsmpte: Int, prevent_dropframe: Boolean): String {
            val ff = bcd2uint(tcsmpte and 0x3f) // 6-bit hours
            val ss = bcd2uint(tcsmpte shr 8 and 0x7f) // 7-bit minutes
            val mm = bcd2uint(tcsmpte shr 16 and 0x7f) // 7-bit seconds
            val hh = bcd2uint(tcsmpte shr 24 and 0x3f) // 6-bit frames
            val drop = tcsmpte and 1 shl 30 > 0L && !prevent_dropframe // 1-bit drop if not arbitrary bit
            return StringUtils.zeroPad2(hh) + ":" +
                    StringUtils.zeroPad2(mm) + ":" +
                    StringUtils.zeroPad2(ss) + (if (drop) ";" else ":") +
                    StringUtils.zeroPad2(ff)
        }

        private fun bcd2uint(bcd: Int): Int {
            val low = bcd and 0xf
            val high = bcd shr 4
            return if (low > 9 || high > 9) 0 else low + 10 * high
        }
    }
}